doxygen/man/man3/tripprogress.3
.TH "tripprogress" 3 "Mon May 11 2020" "Version 1" "CS1D - Project2 - Baseball" \" -*- nroff -*-
.ad l
.nh
.SH NAME
tripprogress \- class for displaying and calculating student trip progress, destinations, and student shopping cart for souverniers  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <tripprogress\&.h>\fP
.PP
Inherits QDialog\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBonLoadClick\fP ()"
.br
.RI "initializes tripProgress on execution , binds tripID, updates size, and calculates nextTrip "
.ti -1c
.RI "void \fBonNextClick\fP ()"
.br
.RI "executes \fBnextTrip()\fP; "
.ti -1c
.RI "void \fBonPrevClick\fP ()"
.br
.RI "executes \fBprevTrip()\fP; "
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBtripprogress\fP (QWidget *parent=nullptr)"
.br
.RI "default explicit Constructor "
.ti -1c
.RI "\fBtripprogress\fP (QWidget *parent, QString I_D)"
.br
.ti -1c
.RI "\fB~tripprogress\fP ()"
.br
.RI "Destructor\&. "
.ti -1c
.RI "void \fBdisplaySouv\fP ()"
.br
.RI "updates sql query to update requested college souverniers and displays \fBUi\fP table with updated model "
.ti -1c
.RI "void \fBnextTrip\fP ()"
.br
.RI "calculates and displays next destination for student planned trip "
.ti -1c
.RI "void \fBprevTrip\fP ()"
.br
.RI "calculates and displays previous destination for student planned trip "
.ti -1c
.RI "void \fBgetDistance\fP ()"
.br
.RI "Retreives teh distance to the next stadium from the database for display\&. "
.ti -1c
.RI "void \fBupdateCart\fP ()"
.br
.RI "calculates and displays updated student cart for souveniers, total and quantity requested by the student at each school destination "
.ti -1c
.RI "void \fBupdateTotal\fP ()"
.br
.RI "calculates and displays updated student price total for shopping souverniers from cart "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "QString \fBtripID\fP"
.br
.RI "field member for storing tripID "
.ti -1c
.RI "QString \fBcurrentCol\fP"
.br
.RI "field member for storing currentCol(db) "
.ti -1c
.RI "QString \fBsouvenirName\fP"
.br
.RI "field member for storing college souvenirName "
.ti -1c
.RI "double \fBsouvenirPrice\fP"
.br
.RI "field member for storing college souvenirPrice "
.ti -1c
.RI "int \fBrow\fP"
.br
.RI "field member for storing sql table row "
.in -1c
.SS "Private Slots"

.in +1c
.ti -1c
.RI "void \fBon_souvTable_clicked\fP (const QModelIndex &index)"
.br
.RI "Gets the data of the chosen cell in the table\&. "
.ti -1c
.RI "void \fBon_addCart_clicked\fP ()"
.br
.RI "updates database, cart, and total; "
.ti -1c
.RI "void \fBon_pushButton_4_released\fP ()"
.br
.RI "Checkout the cart to make a purchase\&. "
.ti -1c
.RI "void \fBon_removeCart_clicked\fP ()"
.br
.RI "Removes an item from the cart\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "Ui::tripprogress * \fBui\fP"
.br
.RI "points to tripprogress UI "
.ti -1c
.RI "\fBDbManager\fP \fBmyDb\fP = \fBDbManager\fP(PROJECT_PATH + '/college\&.db')"
.br
.RI "assigns DB to predefined PROJECT_PATH constant "
.ti -1c
.RI "QStringList \fBtrip\fP"
.br
.RI "stores list of each trip/destination data "
.ti -1c
.RI "QString \fBteamName\fP"
.br
.ti -1c
.RI "int \fBcounter\fP"
.br
.ti -1c
.RI "int \fBmax\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
class for displaying and calculating student trip progress, destinations, and student shopping cart for souverniers 

tripprogress displays an overview of what the user has planned for his trip this updates all data from souvenier, collegename, price, quantity, etc for the plan tripped, as well as the student cart, the student next, and previous destinations 
.PP
Definition at line 28 of file tripprogress\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "tripprogress::tripprogress (QWidget * parent = \fCnullptr\fP)\fC [explicit]\fP"

.PP
default explicit Constructor Constructor sets up ui for tripprogress and initializes counter to default 
.PP
\fBParameters\fP
.RS 4
\fI*parent\fP - for QDialog 
.RE
.PP

.PP
Definition at line 4 of file tripprogress\&.cpp\&.
.PP
References counter\&.
.SS "tripprogress::tripprogress (QWidget * parent, QString I_D)"

.PP
Definition at line 17 of file tripprogress\&.cpp\&.
.PP
References counter, and tripprogress()\&.
.PP
Referenced by tripprogress()\&.
.SS "tripprogress::~tripprogress ()"

.PP
Destructor\&. Close the tripprogress UI 
.PP
Definition at line 32 of file tripprogress\&.cpp\&.
.PP
References ui\&.
.SH "Member Function Documentation"
.PP 
.SS "void tripprogress::displaySouv ()"

.PP
updates sql query to update requested college souverniers and displays \fBUi\fP table with updated model 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 223 of file tripprogress\&.cpp\&.
.SS "void tripprogress::getDistance ()"

.PP
Retreives teh distance to the next stadium from the database for display\&. 
.PP
Definition at line 194 of file tripprogress\&.cpp\&.
.PP
Referenced by onLoadClick(), and onNextClick()\&.
.SS "void tripprogress::nextTrip ()"

.PP
calculates and displays next destination for student planned trip 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 80 of file tripprogress\&.cpp\&.
.PP
References counter, confirmpage::getData(), max, confirmpage::on_okButton_released(), tripSummary::on_school_comboBox_currentIndexChanged(), updateCart(), and updateTotal()\&.
.PP
Referenced by onLoadClick(), and onNextClick()\&.
.SS "void tripprogress::on_addCart_clicked ()\fC [private]\fP, \fC [slot]\fP"

.PP
updates database, cart, and total; 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 323 of file tripprogress\&.cpp\&.
.PP
References updateCart(), and updateTotal()\&.
.SS "void tripprogress::on_pushButton_4_released ()\fC [private]\fP, \fC [slot]\fP"

.PP
Checkout the cart to make a purchase\&. 
.PP
Definition at line 346 of file tripprogress\&.cpp\&.
.PP
References confirmpage::getData(), and confirmpage::on_okButton_released()\&.
.SS "void tripprogress::on_removeCart_clicked ()\fC [private]\fP, \fC [slot]\fP"

.PP
Removes an item from the cart\&. 
.PP
Definition at line 363 of file tripprogress\&.cpp\&.
.PP
References updateCart(), and updateTotal()\&.
.SS "void tripprogress::on_souvTable_clicked (const QModelIndex & index)\fC [private]\fP, \fC [slot]\fP"

.PP
Gets the data of the chosen cell in the table\&. 
.PP
\fBParameters\fP
.RS 4
\fIindex\fP - Cell click in the table 
.RE
.PP

.PP
Definition at line 303 of file tripprogress\&.cpp\&.
.SS "void tripprogress::onLoadClick ()\fC [slot]\fP"

.PP
initializes tripProgress on execution , binds tripID, updates size, and calculates nextTrip 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 39 of file tripprogress\&.cpp\&.
.PP
References getDistance(), and nextTrip()\&.
.SS "void tripprogress::onNextClick ()\fC [slot]\fP"

.PP
executes \fBnextTrip()\fP; 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 246 of file tripprogress\&.cpp\&.
.PP
References getDistance(), and nextTrip()\&.
.SS "void tripprogress::onPrevClick ()\fC [slot]\fP"

.PP
executes \fBprevTrip()\fP; 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 252 of file tripprogress\&.cpp\&.
.PP
References prevTrip()\&.
.SS "void tripprogress::prevTrip ()"

.PP
calculates and displays previous destination for student planned trip 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 156 of file tripprogress\&.cpp\&.
.PP
References counter\&.
.PP
Referenced by onPrevClick()\&.
.SS "void tripprogress::updateCart ()"

.PP
calculates and displays updated student cart for souveniers, total and quantity requested by the student at each school destination 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 257 of file tripprogress\&.cpp\&.
.PP
Referenced by nextTrip(), on_addCart_clicked(), and on_removeCart_clicked()\&.
.SS "void tripprogress::updateTotal ()"

.PP
calculates and displays updated student price total for shopping souverniers from cart 
.PP
\fBReturns\fP
.RS 4
nothing - void 
.RE
.PP

.PP
Definition at line 278 of file tripprogress\&.cpp\&.
.PP
Referenced by nextTrip(), on_addCart_clicked(), and on_removeCart_clicked()\&.
.SH "Member Data Documentation"
.PP 
.SS "int tripprogress::counter\fC [private]\fP"

.PP
Definition at line 173 of file tripprogress\&.h\&.
.PP
Referenced by nextTrip(), prevTrip(), and tripprogress()\&.
.SS "QString tripprogress::currentCol"

.PP
field member for storing currentCol(db) 
.PP
Definition at line 60 of file tripprogress\&.h\&.
.SS "int tripprogress::max\fC [private]\fP"

.PP
Definition at line 174 of file tripprogress\&.h\&.
.PP
Referenced by nextTrip()\&.
.SS "\fBDbManager\fP tripprogress::myDb = \fBDbManager\fP(PROJECT_PATH + '/college\&.db')\fC [private]\fP"

.PP
assigns DB to predefined PROJECT_PATH constant 
.PP
Definition at line 164 of file tripprogress\&.h\&.
.SS "int tripprogress::row"

.PP
field member for storing sql table row 
.PP
Definition at line 75 of file tripprogress\&.h\&.
.SS "QString tripprogress::souvenirName"

.PP
field member for storing college souvenirName 
.PP
Definition at line 65 of file tripprogress\&.h\&.
.SS "double tripprogress::souvenirPrice"

.PP
field member for storing college souvenirPrice 
.PP
Definition at line 70 of file tripprogress\&.h\&.
.SS "QString tripprogress::teamName\fC [private]\fP"

.PP
Definition at line 171 of file tripprogress\&.h\&.
.SS "QStringList tripprogress::trip\fC [private]\fP"

.PP
stores list of each trip/destination data 
.PP
Definition at line 169 of file tripprogress\&.h\&.
.SS "QString tripprogress::tripID"

.PP
field member for storing tripID 
.PP
Definition at line 55 of file tripprogress\&.h\&.
.SS "Ui::tripprogress* tripprogress::ui\fC [private]\fP"

.PP
points to tripprogress UI 
.PP
Definition at line 159 of file tripprogress\&.h\&.
.PP
Referenced by ~tripprogress()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for CS1D - Project2 - Baseball from the source code\&.


tripprogress.cpp
#include "tripprogress.h"
#include "ui_tripprogress.h"
#include <QGraphicsView>
#include <QGraphicsScene>
#include <QMovie>

tripprogress::tripprogress(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::tripprogress)
{
    ui->setupUi(this);
    counter = -1;

    //Hiding previous
    ui->pushButton_2->hide();
    myDb.resetCart();

}

tripprogress::tripprogress(QWidget *parent, QString I_D)
    :QDialog(parent),
      ui(new Ui::tripprogress)
{
    tripID = I_D;

    ui->setupUi(this);
    counter = -1;

    //Hiding previous
    ui->pushButton_2->hide();
    myDb.resetCart();

}

tripprogress::~tripprogress()
{
    myDb.resetCart();

    delete ui;
}

void tripprogress::onLoadClick()
{
    QSqlQuery* qry=new QSqlQuery();

    qry->prepare("SELECT stadium FROM Trips WHERE tripID = (:tripId)");
    qry->bindValue(":tripId" , tripID);

    trip.clear();

    if(!qry->exec())
    {
        qDebug() << "tripProgress initializeList query failed";


    }
    else
    {
        while(qry->next())
        {
            QString temp = qry->value("stadium").toString();
            trip << temp;
        }

    }
    qDebug() << trip;

    max = trip.size();

    nextTrip();

    this->ui->load->deleteLater();

    ui->removeCart->setDisabled(false);
    ui->pushButton_4->setDisabled(false);
    ui->pushButton_2->setDisabled(false);
    ui->pushButton->setDisabled(false);


    getDistance();
}

void tripprogress::nextTrip()// shows next college in trip
{
    ui->addCart->setEnabled(false);
    myDb.resetCart();
    updateCart();
    updateTotal();

    QSqlQueryModel* model=new QSqlQueryModel();

    QSqlQuery* qry=new QSqlQuery();
    qDebug() << counter;
    counter++;

    if( counter < max)
    {
        qry->prepare("SELECT TeamName FROM MLB WHERE StadiumName = (:stadiumName)");
        qry->bindValue(":stadiumName", trip[counter]);

        if(qry->exec())
        {
            qry->first();
            teamName = qry->value(0).toString();
        }

        QString curCol;

        curCol = trip[counter];

        currentCol = trip[counter];


        this->ui->label->setText("CURRENTLY AT: " + curCol);

        qry->prepare("SELECT SouvenirName, Price FROM Souvenirs WHERE TeamName = (:teamName)");
        qry->bindValue(":teamName", teamName);

        if(qry->exec())
        {
            qry->first();
            qDebug() << "souvenir table updated.";
        }

        model->setQuery(*qry);

        ui->souvTable->setModel(model);
        ui->souvTable->resizeColumnsToContents();

    }
    else
    {
        confirmpage confirm;
        bool check = false;

        confirm.setModal(true);
        confirm.exec();
        check = confirm.getData();

        if(check)
        {
            this->hide();
            tripSummary *summary = new tripSummary(tripID, this);

            summary->show();
        }
        else
        {
            counter--;
        }


    }


}


void tripprogress::prevTrip() // shows prev college in trip
{
    QSqlQueryModel* model = new QSqlQueryModel();

    QSqlQuery* qry=new QSqlQuery();

    counter--;

    if( counter >= 0)
    {

        QString curCol;

        curCol = trip[counter];
        currentCol = trip[counter];

        this->ui->label->setText("CURRENTLY AT: " + curCol);

        qry->prepare("SELECT SouvenirName, Price FROM Souvenirs WHERE TeamName = (:teamName)");
        qry->bindValue(":teamName", curCol);

        if(qry->exec())
        {
            qDebug() << "souvenir table updated.";
        }

        model->setQuery(*qry);

        ui->souvTable->setModel(model);
        ui->souvTable->resizeColumnsToContents();
    }
    else
    {
        counter++;
    }

}

void tripprogress::getDistance()
{
    QString distance;

    QSqlQuery* qry=new QSqlQuery();

    qry->prepare("SELECT distanceToNext From Trips WHERE (stadium) = (:stadium)");
    qry->bindValue(":stadium", currentCol);

    if(qry->exec())
    {
        while(qry->next())
        {
            distance = qry->value(0).toString();
        }

        if(counter+1 < max)
            distance = "Distance to " + trip[counter+1] + ": " + distance;
        else
            distance = "You're at the last stadium!";

        ui->distanceLabel->setText(distance);
        qDebug() << distance;
        qDebug() << "show next college!";
    }
    else
        qDebug() << "doesn't show next colelge!";
}

void tripprogress::displaySouv()
{
    QSqlQueryModel* model = new QSqlQueryModel();

    QSqlQuery* qry=new QSqlQuery();

    qry->prepare("SELECT SouvenirName FROM Souvenirs WHERE StadiumName = (:stadiumName)");
    qry->bindValue(":stadiumName", currentCol);

    if(qry->exec())
    {
        qDebug() << "college table updated.";
    }
    else
        qDebug() << "failed";

    model->setQuery(*qry);

    ui->souvTable->setModel(model);
    ui->souvTable->setColumnWidth(20, 400);

}

void tripprogress::onNextClick()
{
    nextTrip();
    getDistance();
}

void tripprogress::onPrevClick()
{
    prevTrip();
}

void tripprogress::updateCart()
{

    QSqlQueryModel* model=new QSqlQueryModel();

    QSqlQuery* qry=new QSqlQuery();

    qry->prepare("SELECT souvenir, quantity, total FROM Cart");

    if(qry->exec())
    {
        qDebug() << "college1 table updated.";
    }
    else
        qDebug() << "failed";

    model->setQuery(*qry);

    ui->cartTable->setModel(model);
}

void tripprogress::updateTotal()
{
    QSqlQueryModel* model=new QSqlQueryModel();

    QSqlQuery* qry=new QSqlQuery();

    qry->prepare("SELECT SUM(total) FROM Cart");

    if(qry->exec())
    {
        qDebug() << "college1 table updated.";
    }
    else
        qDebug() << "failed";

    model->setQuery(*qry);

    model->setHeaderData(0, Qt::Horizontal, tr("Cart Total") );

    ui->totalView->setModel(model);




}
void tripprogress::on_souvTable_clicked(const QModelIndex &index)
{

    if(index.isValid())
    {
        ui->addCart->setEnabled(true);
        row = index.row();
        souvenirName = index.sibling(row, 0).data().toString();
        souvenirPrice = index.sibling(row, 1).data().toDouble();

        qDebug() << endl;

        qDebug() << tripID << " " << currentCol;
        qDebug() << souvenirName << " " << souvenirPrice << endl;
        qDebug() << index << endl;

    }

}

void tripprogress::on_addCart_clicked()
{
    int count = 1;
    int counter = myDb.cartQuantity(currentCol, souvenirName);

    if(counter == 0)
    {
        myDb.addCart(tripID,currentCol,souvenirName, souvenirPrice,count);
    }
    else
    {
        counter++;


        myDb.updateCart(currentCol,souvenirName,counter);
        qDebug() << counter;

    }
    updateTotal();
    updateCart();
}


void tripprogress::on_pushButton_4_released()
{
    confirmpage confirm;
    bool check;

    confirm.setModal(true);
    confirm.exec();
    check = confirm.getData();

    myDb.reOpen();
    if(check)
    {
        myDb.addPurchase();
        myDb.resetCart();
    }
}

void tripprogress::on_removeCart_clicked()
{
    myDb.removeCart(currentCol, souvenirName);
    updateCart();
    updateTotal();
}
